aquacomputer_d5next.c

```diff
@@ -124,16 +208,420 @@ struct d5next_data {
+    int ret;
+    u16 checksum = 0xffff; /* Init value for CRC-16/USB */
+    checksum = crc16(checksum, (u8*)(priv->buffer) + STATUS_REPORT_CHECKSUM_START, STATUS_REPORT_CHECKSUM_LENGTH);
+    checksum ^= 0xffff; /* Xorout value for CRC-16/USB */
+
+    /* Place the new checksum at the end of the report */
+    put_unaligned_be16(checksum, &(priv->buffer->crc));
+
+    /* Send the patched up report back to the pump */
+    ret = hid_hw_raw_request(priv->hdev, STATUS_REPORT_ID, (u8*)(priv->buffer), STATUS_REPORT_SIZE, HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
+    if (ret < 0)
+        goto exit;
+
+    /* The official software sends this report after every change, so do it here as well */
+    ret = hid_hw_raw_request(priv->hdev, SECONDARY_STATUS_REPORT_ID, secondary_status_report, SECONDARY_STATUS_REPORT_SIZE, HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
+    if (ret < 0)
+        goto exit;
+
+    exit:
+    return ret;
+}
+
+/*
+ * Don't call this directly -- it's for d5next_set_u16 and d5next_set_u8 to call for you
+ */
+static int d5next_set_val(struct device *dev, void* to_set, long val, size_t size)
+{
+    int ret;
+    struct d5next_data *priv = dev_get_drvdata(dev);
+
+    mutex_lock(&priv->mutex);
+
+    /*
+     * Request a complete config report from the pump by sending a GET_FEATURE_REPORT and
+     * store it in dev->buffer
+    */
+    ret = get_d5_ctrl_data(dev);
+    if (ret < 0)
+        goto unlock_and_return;
+
+    /*
+     * Set values accordingly. We only modify those values so that any other
+     * settings, such as RGB lights, stay untouched
+     */
+    switch(size) {
+        case 16:
+            put_unaligned_be16(val, to_set);
+            break;
+        case 8:
+            *(u8 *)to_set = val;
+            break;
+        default:
+            return -EINVAL;
+    }
+
+
+    ret = d5next_tx_buf(priv);
+
+    unlock_and_return:
+    mutex_unlock(&priv->mutex);
+    return ret;
+}
+
+static int d5next_set_u16(struct device *dev, u16* to_set, long val) {
+    return d5next_set_val(dev, to_set, val, 16);
+}
+
+static int d5next_set_u8(struct device *dev, u8* to_set, long val) {
+    return d5next_set_val(dev, to_set, val, 8);
+}
+
+static bool fan_channel_is_valid(int channel) {
+    return (channel >=0 && channel < FAN_CONTROL_CHANNEL_MAX);
+}
+
+static int d5_pwm_percent_to_256(u16 x) {
+    return DIV_ROUND_CLOSEST(ntohs(x)*255, 100 * 100);
+}
+
+static ssize_t get_pwm_setting(struct device *dev, struct device_attribute *attr, char *buf) {
+    struct sensor_device_attribute_2 *sensor_attr = to_sensor_dev_attr_2(attr);
+    int channel = sensor_attr->nr;
+    int idx = sensor_attr->index;
+    struct d5next_data *priv = dev_get_drvdata(dev);
+    int ret;
+    int val;
+    mutex_lock(&priv->mutex);
+    ret = get_d5_ctrl_data(dev);
+
+    if (ret < 0 || !fan_channel_is_valid(channel)) {
+        ret = -ENODATA;
+        goto exit;
+    }
+    else if (idx == IDX_MIN_PWM)
```

Comment:
Are you sure? I wasn't wild about a switch with 16 identical branches.

Reply:
I'm only saying that because most of the hwmon code I saw was using switches for this type of stuff. If statements are generally fine by me, just saying it could be a switch since it depends on a single variable